flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sortpartition/FixedLengthByteKeyAndValueComparator.java

```diff
@@ -0,0 +1,176 @@
+ * be fixed and will be initialized when the {@link FixedLengthByteKeyAndValueComparator} is
+ * created.
+ */
+@Internal
+public class FixedLengthByteKeyAndValueComparator<INPUT>
+        extends TypeComparator<Tuple2<byte[], INPUT>> {
+
+    private final int serializedKeyLength;
+
+    private final TypeComparator<INPUT> valueComparator;
+
+    private byte[] keyReference;
+
+    private INPUT valueReference;
+
+    FixedLengthByteKeyAndValueComparator(
+            int serializedKeyLength, TypeComparator<INPUT> valueComparator) {
+        this.serializedKeyLength = serializedKeyLength;
+        this.valueComparator = valueComparator;
+    }
+
+    @Override
+    public int hash(Tuple2<byte[], INPUT> record) {
+        return record.hashCode();
+    }
+
+    @Override
+    public void setReference(Tuple2<byte[], INPUT> toCompare) {
+        this.keyReference = toCompare.f0;
+        this.valueReference = toCompare.f1;
+    }
+
+    @Override
+    public boolean equalToReference(Tuple2<byte[], INPUT> candidate) {
+        return Arrays.equals(keyReference, candidate.f0) && valueReference == candidate.f1;
+    }
+
+    @Override
+    public int compareToReference(TypeComparator<Tuple2<byte[], INPUT>> referencedComparator) {
+        byte[] otherKey =
+                ((FixedLengthByteKeyAndValueComparator<INPUT>) referencedComparator).keyReference;
+        INPUT otherValue =
+                ((FixedLengthByteKeyAndValueComparator<INPUT>) referencedComparator).valueReference;
+        int keyCmp = compare(otherKey, this.keyReference);
+        if (keyCmp != 0) {
+            return keyCmp;
+        }
+        return valueComparator.compare(this.valueReference, otherValue);
+    }
+
+    @Override
+    public int compare(Tuple2<byte[], INPUT> first, Tuple2<byte[], INPUT> second) {
+        int keyCmp = compare(first.f0, second.f0);
+        if (keyCmp != 0) {
+            return keyCmp;
+        }
+        return valueComparator.compare(first.f1, second.f1);
+    }
+
+    private int compare(byte[] first, byte[] second) {
+        for (int i = 0; i < serializedKeyLength; i++) {
+            int cmp = Byte.compare(first[i], second[i]);
+            if (cmp != 0) {
+                return cmp < 0 ? -1 : 1;
+            }
+        }
+        return 0;
+    }
+
+    @Override
+    public int compareSerialized(DataInputView firstSource, DataInputView secondSource)
+            throws IOException {
+        int minCount = serializedKeyLength;
+        while (minCount-- > 0) {
+            byte firstValue = firstSource.readByte();
+            byte secondValue = secondSource.readByte();
+            int cmp = Byte.compare(firstValue, secondValue);
+            if (cmp != 0) {
+                return cmp < 0 ? -1 : 1;
+            }
+        }
+        return valueComparator.compareSerialized(firstSource, secondSource);
+    }
+
+    @Override
+    public boolean supportsNormalizedKey() {
```

Comment:
`NormalizedKey` is used to map data to a standardized representation that can be compared, which is a byte array. `NormalizedKey` is used by `NormalizedKeySorter` for sorting in memory. In the implementation of the `SortPartition` API, we use `ExternalSorter` for external sorting. BTW, it's hard for each piece of sorted data to generate a `NormalizedKey` for comparison since it contains both key and value with various types .